### ISSUE TITLE
Time stamp offset

### ISSUE THREAD

**Issue by AndreasAZiegler** [ros-event-camera/event_camera_codecs#12]
During recent work with [FrequencyCam](https://github.com/berndpfrommer/frequency_cam) where we used the trigger-in events to synchronize the output of FrequencyCam with the output of a frame-based processing pipeline, we noticed a time stamp offset.

When we adjusted `evt3_decoder.h` from
```
inline timestamp_t makeTime(timestamp_t high, uint16_t low) { return ((high | low) * timeMult_); }
```
to
```
inline timestamp_t makeTime(timestamp_t high, uint16_t low) {
  timestamp_t ts = (high | low) * timeMult_;
  ts -= 16777215000;
  return (ts);
}
```
the time stamp offset was gone.

We used [this ROS2 bag](https://cloud.cs.uni-tuebingen.de/index.php/s/diy5YrXJofDWMZa) which is generated from a `.raw` file.

**Comment by berndpfrommer** [comment]
An offset is always between two different time stamps. The question is now: how are the two different time stamps generated? One of them seems to be coming from the decoder by virtue of "makeTime()". That part is clear. But what about the other timestamp ("output of a frame-based processing pipeline")?
I have to guess that both time stamps are computed from the sensor time stamps embedded in an evt3 encoded event stream, so both use makeTime(). Then to get an offset between sensor times, the sensor times either come from different event cameras, or  the decoding of the evt3 stream has started at a different points in the stream. Note that the "high" time stamp in the event stream rolls over every 16.8 seconds, exactly the offset you are seeing. So  let's say you start decoding at time t in program A, but at time t+epsilon in program B, and the "high" timestamp just so happens to roll over between t and t + epsilon. Then program B will see a sensor time that is *smaller* by 16.8s because program A observes the rollover, and thus adds 16.8s to the time stamp, whereas program B does not.

**Comment by AndreasAZiegler** [comment]
To record the frames and the event streams, we have a tool which uses the Spinnaker library to save the frames and Metavision to save the streams as `.raw` files. We don't really synchronize the time stamps. We use an external trigger source (a microcontroller) which triggers the frame-based cameras to capture a frame (with a defined exposure time (5 ms)) and generate trigger-in events in the main event-based camera (the two event-based cameras are synchronized via a sync cable).
We extracted the time stamps (trigger-in events plus 5ms exposure time) so that FrequencyCam generates the output at the same time the frame-based cameras finished the exposure time. By visually comparing the frames and the output of FrequencyCam, we noticed a time mismatch. Being aware of the 16.8 seconds, I used this number as "correction" term in the decoder, which led to matching outputs. I hope my explanation was clear, otherwise, please let me know. I tend to be too much used to our setup and might fail to explain it to third parties.

**Comment by berndpfrommer** [comment]
Do I understand this right:
event_stream_event_camera_1 ---(decode starting from t_1, extract trigger-in events)-->  time_stamp_1
event_stream_event_camera _2 --(decode starting from t_2, extract cd events up to time_stamp_1)--> frequency_cam --> image

So when you do this, you see that the images generated by frequency_cam are off by 16.8s w.r.t to the frame based camera images?

Questions:
1) Are event_stream_camera_1 and event_stream_camera_2 identical, i.e. are the trigger events coming from the same camera that you later use to run frequency cam?
2) If they are from the same camera, do you start decoding from the very start in both cases?
3) how do you synchronize between sensor time and ros time stamps to get the association between frame based camera and frequency_cam frames? I assume you use event_stream_camera_1, i.e. the stream with the trigger events, and there you look at the offset between sensor time and ros header time stamp?

**Comment by AndreasAZiegler** [comment]
It looks like you got our pipeline right. However, to make sure, I will present it slightly different.
event_stream_event_camera_1.raw --> time_stamps
event_stream_event_camera_1.raw --> event_stream_event_camera_1 (ROS2 bag)
event_stream_event_camera_2.raw --> event_stream_event_camera_2 (ROS2 bag)

event_stream_event_camera_1 (ROS2 bag), time_stamps --> frequency_cam --> frequency_images_1
event_stream_event_camera_2 (ROS2 bag), time_stamps --> frequency_cam --> frequency_images_2

You are right, the images generated by frequency_cam are off w.r.t. the frame based camera images.

Now to your questions:
1. Only the main/master event-based cameras receives trigger-in signals. But given that the two event-based cameras are connected with the 1MHz sync signal, they should be synchronized among themselfes
2. We use the `playEventsFromBag` method from `FrequencyCamROS` for both ROS2 bags, therefore, I assume that the decoding starts from the very start. Given the fact that the frequency_cam output of both event-streams is off w.r.t. the frame based images, the decoding seems the same for both ROS2 bags.
3. We don't use the ROS time stamps. From my understanding, frequency_cam only use the time stamps in `frameTimes_` and `nextTime` also represents sensor time stamps. We do the association between the FB cameras and the EB cameras simply by the assumption that the first FB frames should match to the frequency_cam output at the first extracted time stamp.

**Comment by berndpfrommer** [comment]
Good, in this case we can ignore camera 2 completely since the problem shows up already for camera 1.
What software do you use to decode the time stamps? Is that also using event_camera_codecs, or some other decoder?

**Comment by AndreasAZiegler** [comment]
I used [this python script](https://github.com/AndreasAZiegler/frequency_cam/blob/csv-output/scripts/extract_triggers_prophesee.py). It is an adjusted version of [this script](https://github.com/uzh-rpg/e2calib/blob/main/python/extract_triggers_prophesee.py) also using `Metavision`.

**Comment by berndpfrommer** [comment]
Well, you are using the Metavision decoder then. The moment you bring in a different decoder, all bets are off. At some point in the distant past I tried to understand the time stamps that their iterator produces, but eventually I gave up. The metavision toolkit offers two ways to fiddle with the time stamps, one at the [RawReader](https://docs.prophesee.ai/stable/api/python/core/event_io.html#metavision_core.event_io.RawReader)  level, the other at the [EventIterator](https://docs.prophesee.ai/stable/api/python/core/event_io.html#metavision_core.event_io.EventsIterator) level ("relative_timestamps"). I played around with those options but found the time stamps it produced highly confusing, and often shifted with respect to what my decoder produced.
I wish the decoding logic for the "high" time stamp was cleaner, but the handling of bit errors makes the code hard to read. Still, I tested it extensively and it produced results in agreement with the metavision decoder, up to some fixed shift.
